CODE_STYLE.md

```diff
@@ -0,0 +1,3293 @@
+  blocks' contents be extracted into nifty methods?
+
+* <a name="block-argument"></a>
+  Consider using explicit block argument to avoid writing block literal that
+  just passes its arguments to another block. Beware of the performance impact,
+  though, as the block gets converted to a Proc.
+<sup>[[link](#block-argument)]</sup>
+
+  ```Ruby
+  require 'tempfile'
+
+  # bad
+  def with_tmp_dir
+    Dir.mktmpdir do |tmp_dir|
+      Dir.chdir(tmp_dir) { |dir| yield dir }  # block just passes arguments
+    end
+  end
+
+  # good
+  def with_tmp_dir(&block)
+    Dir.mktmpdir do |tmp_dir|
+      Dir.chdir(tmp_dir, &block)
+    end
+  end
+
+  with_tmp_dir do |dir|
+    puts "dir is accessible as a parameter and pwd is set: #{dir}"
+  end
+  ```
+
+* <a name="no-explicit-return"></a>
+  Avoid `return` where not required for flow of control.
+<sup>[[link](#no-explicit-return)]</sup>
+
+  ```Ruby
+  # bad
+  def some_method(some_arr)
+    return some_arr.size
+  end
+
+  # good
+  def some_method(some_arr)
+    some_arr.size
+  end
+  ```
+
+* <a name="no-self-unless-required"></a>
+  Avoid `self` where not required. (It is only required when calling a self
+  write accessor.)
+<sup>[[link](#no-self-unless-required)]</sup>
+
+  ```Ruby
+  # bad
+  def ready?
+    if self.last_reviewed_at > self.last_updated_at
+      self.worker.update(self.content, self.options)
+      self.status = :in_progress
+    end
+    self.status == :verified
+  end
+
+  # good
+  def ready?
+    if last_reviewed_at > last_updated_at
+      worker.update(content, options)
+      self.status = :in_progress
+    end
+    status == :verified
+  end
+  ```
+
+* <a name="no-shadowing"></a>
+  As a corollary, avoid shadowing methods with local variables unless they are
+  both equivalent.
+<sup>[[link](#no-shadowing)]</sup>
+
+  ```Ruby
+  class Foo
+    attr_accessor :options
+
+    # ok
+    def initialize(options)
+      self.options = options
+      # both options and self.options are equivalent here
+    end
+
+    # bad
+    def do_something(options = {})
+      unless options[:when] == :later
+        output(self.options[:message])
+      end
+    end
+
+    # good
+    def do_something(params = {})
+      unless params[:when] == :later
+        output(options[:message])
+      end
+    end
+  end
+  ```
+
+* <a name="safe-assignment-in-condition"></a>
+  Don't use the return value of `=` (an assignment) in conditional expressions
+  unless the assignment is wrapped in parentheses. This is a fairly popular
+  idiom among Rubyists that's sometimes referred to as *safe assignment in
+  condition*.
+<sup>[[link](#safe-assignment-in-condition)]</sup>
+
+  ```Ruby
+  # bad (+ a warning)
+  if v = array.grep(/foo/)
+    do_something(v)
+    ...
+  end
+
+  # good (MRI would still complain, but RuboCop won't)
+  if (v = array.grep(/foo/))
+    do_something(v)
+    ...
+  end
+
+  # good
+  v = array.grep(/foo/)
+  if v
+    do_something(v)
+    ...
+  end
```

Comment:
Same. Both C and Go seem to encourage this kind of thing, but i feel most of the time it obfuscates things. Checking a condition _and_ assigning a variable in the same line - too much at once - visually.